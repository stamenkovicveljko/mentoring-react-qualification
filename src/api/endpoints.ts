const githubUserEndpoint = (user: string) =>
  `https://api.github.com/users/${user}`;

const githubRepoEndpoint = (user: string) =>
  `https://api.github.com/users/${user}/repos`;

export default {
  githubRepoEndpoint,
  githubUserEndpoint,
};
