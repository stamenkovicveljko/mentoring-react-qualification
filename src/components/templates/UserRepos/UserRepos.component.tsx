import React, { useState } from "react";
import IGithubUser from "../../../models/GithubUser";
import IGithubRepo from "../../../models/GithubRepo";
import ProfileInfo from "../../organisms/ProfileInfo/ProfileInfo.component";
import Input from "../../atoms/Input/Input.component";
import List from "../../molecules/List/List.component";
import Link from "../../atoms/Link/Link.component";
import * as Styled from "./UserRepos.styles";

interface Props {
  user: IGithubUser;
  repos: IGithubRepo[];
}

const UserRepos = ({ user, repos }: Props) => {
  const [term, setTerm] = useState<string>("");

  const handleTermOnChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
    setTerm(ev.target.value);
  };

  return (
    <Styled.Container>
      <ProfileInfo user={user} className="profile-info" />
      <div>
        <Input
          value={term}
          onChange={handleTermOnChange}
          placeholder="Search repos"
        />
        <List
          items={repos
            .filter(({ name }) =>
              name.toLowerCase().includes(term.toLowerCase())
            )
            .map((repo) => (
              <Link to={repo.html_url}>{repo.name}</Link>
            ))}
        />
      </div>
    </Styled.Container>
  );
};
export default UserRepos;
