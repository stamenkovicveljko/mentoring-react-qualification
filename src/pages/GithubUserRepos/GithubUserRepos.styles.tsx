import styled from "styled-components";

export const Container = styled.div`
  max-width: 768px;
  margin: 0 auto;
  padding: 24px;
`;

export const UserSearch = styled.div`
  display: flex;
  margin-bottom: 16px;

  > div {
    flex: 1 1 100%;
  }

  label {
    display: block;
    margin-bottom: 8px;
  }

  button {
    flex: 0 0 auto;
    margin-left: 16px;
    align-self: flex-end;
  }
`;
