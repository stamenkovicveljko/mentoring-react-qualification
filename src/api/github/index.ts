import endpoints from "../endpoints";
import { reqOptions, handleErrors } from "..";

function getUser(username: string) {
  return fetch(endpoints.githubUserEndpoint(username), reqOptions)
    .then(handleErrors)
    .then((res) => res.json());
}

function getRepos(username: string) {
  return fetch(endpoints.githubRepoEndpoint(username), reqOptions)
    .then(handleErrors)
    .then((res) => res.json());
}

export default {
  getUser,
  getRepos,
};
