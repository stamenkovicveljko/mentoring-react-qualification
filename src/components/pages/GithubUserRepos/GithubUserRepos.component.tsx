import React, { ReactElement, useState, useEffect } from "react";
import IGithubRepo from "../../../models/GithubRepo";
import IGithubUser from "../../../models/GithubUser";
import endpoints from "../../../api/endpoints";
import UserRepos from "../../templates/UserRepos/UserRepos.component";
import ErrorTemplate from "../../templates/Error/Error.component";
import * as Styled from "./GithubUserRepos.styles";
import Input from "../../atoms/Input/Input.component";
import Button from "../../atoms/Button/Button.component";

function handleErrors(response: Response) {
  if (!response.ok) {
    throw Error(response.statusText);
  }
  return response;
}

export default function GithubUserRepos(): ReactElement {
  const [repos, setRepos] = useState<IGithubRepo[]>([]);
  const [loading, setLoading] = useState<boolean>(false);
  const [userTerm, setUserTerm] = useState<string>("stamenkovicveljko");
  const [selectedUsername, setSelectedUsername] = useState<string>(userTerm);
  const [error, setError] = useState<boolean>(false);

  const [currentUser, setCurrentUser] = useState<IGithubUser | null>();

  useEffect(() => {
    const getRepos = async () => {
      setError(false);
      const reqOptions = {
        headers: {
          Accept: "application/json",
        },
      };

      try {
        setLoading(true);
        const [user, repositories] = await Promise.all<
          IGithubUser,
          IGithubRepo[]
        >([
          (
            await fetch(
              endpoints.githubUserEndpoint(selectedUsername),
              reqOptions
            ).then(handleErrors)
          ).json(),
          (
            await fetch(
              endpoints.githubRepoEndpoint(selectedUsername),
              reqOptions
            ).then(handleErrors)
          ).json(),
        ]);

        setCurrentUser(user);
        setRepos(repositories);
      } catch (er) {
        setCurrentUser(null);
        setRepos([]);
        setError(true);
      } finally {
        setLoading(false);
      }
    };

    getRepos();
  }, [selectedUsername]);

  return (
    <Styled.Container>
      {error ? (
        <ErrorTemplate>
          <p>
            Click&nbsp;
            <Button
              onClick={() => {
                setUserTerm("stamenkovicveljko");
                setSelectedUsername("stamenkovicveljko");
              }}
              variant="secondary"
            >
              here
            </Button>
            &nbsp;to return to default profile.
          </p>
        </ErrorTemplate>
      ) : loading || !currentUser ? (
        <p>loading ...</p>
      ) : (
        <>
          <Styled.UserSearch>
            <div>
              <label htmlFor="user-search">Current user:</label>
              <Input
                id="user-search"
                value={userTerm}
                onChange={(ev) => setUserTerm(ev.target.value)}
                placeholder="Search repos"
              />
            </div>
            <Button
              variant="primary"
              onClick={() => setSelectedUsername(userTerm)}
            >
              Search user
            </Button>
          </Styled.UserSearch>
          <UserRepos repos={repos} user={currentUser} />
        </>
      )}
    </Styled.Container>
  );
}
