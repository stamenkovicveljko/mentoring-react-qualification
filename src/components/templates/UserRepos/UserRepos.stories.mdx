import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import UserRepos from "./UserRepos.component";

<Meta title="UserRepos" component={UserRepos} />

# UserRepos Component

Creates searchable list of specified user's repos from the github.

## Implementation

It relies on `user` and `repos` to be defined.

_User model_

```ts
interface IGithubUser {
  name: string;
  login: string;
  avatar_url: string;
  html_url: string;
}
```

_Repo model_

```ts
interface IGithubRepo {
  name: string;
  html_url: string;
}
```

<Preview>
  <Story name="default">
    <UserRepos
      user={{
        name: "Some user",
        login: "storiez",
        avatar_url:
          "https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg",
        html_url:
          "https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg",
      }}
      repos={[
        {
          name: "some repo 1",
          html_url: "https://google.com",
        },
        {
          name: "some repo 2",
          html_url: "https://google.com",
        },
        {
          name: "some repo 3",
          html_url: "https://google.com",
        },
      ]}
    />
  </Story>
</Preview>

## Props

<Props of={UserRepos} />
