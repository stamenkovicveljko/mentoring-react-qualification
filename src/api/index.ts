import github from "./github";

export function handleErrors(response: Response) {
  if (!response.ok) {
    throw Error(response.statusText);
  }
  return response;
}

export const reqOptions = {
  headers: {
    Accept: "application/json",
  },
};

export default {
  github,
};
