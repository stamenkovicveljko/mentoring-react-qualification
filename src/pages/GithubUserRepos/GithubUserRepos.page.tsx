import React, { ReactElement, useState } from "react";
import IGithubRepo from "../../models/GithubRepo";
import IGithubUser from "../../models/GithubUser";
import UserRepos from "../../components/templates/UserRepos/UserRepos.component";
import ErrorTemplate from "../../components/templates/Error/Error.component";
import * as Styled from "./GithubUserRepos.styles";
import Input from "../../components/atoms/Input/Input.component";
import Button from "../../components/atoms/Button/Button.component";
import api from "../../api";

export default function GithubUserRepos(): ReactElement {
  const [repos, setRepos] = useState<IGithubRepo[]>([]);
  const [loading, setLoading] = useState<boolean>(false);
  const [userTerm, setUserTerm] = useState<string>("stamenkovicveljko");
  const [selectedUsername, setSelectedUsername] = useState<string>(userTerm);
  const [error, setError] = useState<boolean>(false);

  const [currentUser, setCurrentUser] = useState<IGithubUser | null>();

  React.useEffect(() => {
    const getRepos = async () => {
      setError(false);
      try {
        setLoading(true);
        const [user, repositories] = await Promise.all<
          IGithubUser,
          IGithubRepo[]
        >([
          api.github.getUser(selectedUsername),
          api.github.getRepos(selectedUsername),
        ]);

        setCurrentUser(user);
        setRepos(repositories);
      } catch (er) {
        setCurrentUser(null);
        setRepos([]);
        setError(true);
      } finally {
        setLoading(false);
      }
    };

    getRepos();
  }, [selectedUsername]);

  return (
    <Styled.Container>
      {error ? (
        <ErrorTemplate>
          <p>
            Click&nbsp;
            <Button
              onClick={() => {
                setUserTerm("stamenkovicveljko");
                setSelectedUsername("stamenkovicveljko");
              }}
              variant="secondary"
            >
              here
            </Button>
            &nbsp;to return to default profile.
          </p>
        </ErrorTemplate>
      ) : loading || !currentUser ? (
        <p>loading ...</p>
      ) : (
        <>
          <Styled.UserSearch>
            <div>
              <label htmlFor="user-search">Current user:</label>
              <Input
                id="user-search"
                value={userTerm}
                onChange={(ev) => setUserTerm(ev.target.value)}
                placeholder="Search repos"
              />
            </div>
            <Button
              variant="primary"
              onClick={() => setSelectedUsername(userTerm)}
            >
              Search user
            </Button>
          </Styled.UserSearch>
          <UserRepos repos={repos} user={currentUser} />
        </>
      )}
    </Styled.Container>
  );
}
