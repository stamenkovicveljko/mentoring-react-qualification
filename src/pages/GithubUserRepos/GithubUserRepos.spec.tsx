import React from "react";
import renderer from "react-test-renderer";
import GithubUserRepos from "./GithubUserRepos.page";
import { shallow, ShallowWrapper } from "enzyme";
import api from "../../api";
import UserRepos from "../../components/templates/UserRepos/UserRepos.component";

let useEffect: jest.SpyInstance<
  void,
  [React.EffectCallback, (React.DependencyList | undefined)?]
>;
let wrapper: ShallowWrapper;
let originalGetUser = api.github.getUser;
let originalGetRepos = api.github.getRepos;

useEffect = jest.spyOn(React, "useEffect");

const mockUseEffect = () => {
  useEffect.mockImplementationOnce((f: () => any) => f());
};

describe("GithubUserRepos page", () => {
  const user = {
    name: "Some user",
    login: "storiez",
    avatar_url:
      "https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg",
    html_url: "https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg",
  };

  const repos = [
    {
      name: "some repo 1",
      html_url: "https://google.com",
    },
    {
      name: "some repo 2",
      html_url: "https://google.com",
    },
    {
      name: "some repo 3",
      html_url: "https://google.com",
    },
  ];

  beforeAll(() => {
    api.github.getUser = jest.fn().mockResolvedValue(user);
    api.github.getRepos = jest.fn().mockResolvedValue(repos);
  });

  afterAll(() => {
    api.github.getUser = originalGetUser;
    api.github.getRepos = originalGetRepos;
  });

  beforeEach(() => {
    mockUseEffect();
    wrapper = shallow(<GithubUserRepos />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test("GithubUserRepos renders as expected", () => {
    const component = renderer.create(<GithubUserRepos />);

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe("on start", () => {
    it("loads user", () => {
      expect(api.github.getUser).toHaveBeenCalled();
    });

    it("loads repos", () => {
      expect(api.github.getRepos).toHaveBeenCalled();
    });
  });

  describe("on render", () => {
    it("renders user repoes", () => {
      expect(wrapper.find(UserRepos).length).toBe(1);
    });
  });
});
