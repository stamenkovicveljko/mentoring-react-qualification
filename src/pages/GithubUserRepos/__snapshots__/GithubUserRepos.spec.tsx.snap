// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubUserRepos page GithubUserRepos renders as expected 1`] = `
<div
  className="sc-AxheI fLZSOK"
>
  <p>
    loading ...
  </p>
</div>
`;
